### PR TITLE
Add .fold() method to Zip

### PR DESCRIPTION
**Pull request by jturner314**
This is just for convenience to avoid having to deal with `FoldWhile`.

### CODE REVIEW

File: src/zip/mod.rs [hunk 1]
```diff
@@ -755,8 +755,44 @@ macro_rules! map_impl {
...
+            /// let sum_prod_zip = Zip::from(&a).and(&b).fold(0, |acc, a, b| acc + a * b);
+            ///
+            /// assert_eq!(sum_prod_nonzip, sum_prod_zip);
+            /// ```
+            pub fn fold<F, Acc>(mut self, acc: Acc, mut function: F) -> Acc
+            where
+                F: FnMut(Acc, $($p::Item),*) -> Acc,
+            {
+                self.apply_core(acc, move |acc, args| {
+                    let ($($p,)*) = args;
```
Comment (max-sixty) [inline src/zip/mod.rs]: @jturner314 for my education: I hadn't seen this syntax outside macros. Do you have any reference to docs explaining what's going on?

Comment (Jvanrhijn) [reply src/zip/mod.rs]: It is inside a macro, starts at line [736](https://github.com/rust-ndarray/ndarray/blob/27c010dbe3b3efc2cdd62e2f25d44da07ef97c8e/src/zip/mod.rs#L736). I had to do a double take there :smile:

Comment (max-sixty) [reply src/zip/mod.rs]: Ah! Thanks :)

**Comment by LukeMathWalker** [comment]
This makes the API nicer to use - I don't see any downside, we can go ahead and merge.

**Comment by jturner314** [comment]
Thanks for reviewing this @LukeMathWalker!

**Pull request merged**